File: apps/demo-be-java-springboot/src/test/java/com/demobejasb/integration/user_account/UserAccountSteps.java
package com.demobejasb.integration.user_account;

import com.demobejasb.auth.repository.UserRepository;
import com.demobejasb.integration.ResponseStore;
import com.demobejasb.integration.steps.TokenStore;
import com.demobejasb.security.JwtUtil;
import com.demobejasb.user.dto.UserProfileResponse;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.When;
import java.util.Map;
import java.util.Objects;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Scope;
import org.springframework.security.crypto.password.PasswordEncoder;

@Scope("cucumber-glue")
public class UserAccountSteps {

    @Autowired
    private ResponseStore responseStore;

    @Autowired
    private TokenStore tokenStore;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    @Autowired
    private JwtUtil jwtUtil;

    @When("^alice sends GET /api/v1/users/me$")
    public void aliceSendsGetUsersMe() {
        String token = tokenStore.getToken();
        if (token == null) {
            throw new IllegalStateException("Token not stored");
        }
        performGetUsersMe(token);
    }

    @When("^alice sends PATCH /api/v1/users/me with body \\{ \"display_name\": \"Alice Smith\" \\}$")
    public void aliceSendsPatchUsersMeWithDisplayName() {
        String token = tokenStore.getToken();
        if (token == null) {
            throw new IllegalStateException("Token not stored");
        }
        if (!jwtUtil.isTokenValid(token)) {
            responseStore.setResponse(401, Map.of("message", "Unauthorized"));
            return;
        }
        String username = jwtUtil.extractUsername(token);
        userRepository.findByUsername(username).ifPresentOrElse(user -> {
            user.setDisplayName("Alice Smith");
            userRepository.save(user);
            responseStore.setResponse(200, UserProfileResponse.from(user));
        }, () -> responseStore.setResponse(404, Map.of("message", "User not found")));
    }

    @When("^alice sends POST /api/v1/users/me/password with body \\{ \"old_password\": \"Str0ng#Pass1\", \"new_password\": \"NewPass#456\" \\}$")
    public void aliceSendsPostChangePasswordSuccess() {
        String token = tokenStore.getToken();
        if (token == null) {
            throw new IllegalStateException("Token not stored");
        }
        performChangePassword(token, "Str0ng#Pass1", "NewPass#456");
    }

    @When("^alice sends POST /api/v1/users/me/password with body \\{ \"old_password\": \"Wr0ngOld!\", \"new_password\": \"NewPass#456\" \\}$")
    public void aliceSendsPostChangePasswordWrongOld() {
        String token = tokenStore.getToken();
        if (token == null) {
            throw new IllegalStateException("Token not stored");
        }
        performChangePassword(token, "Wr0ngOld!", "NewPass#456");
    }

    @When("^alice sends POST /api/v1/users/me/deactivate$")
    public void aliceSendsPostSelfDeactivate() {
        String token = tokenStore.getToken();
        if (token == null) {
            throw new IllegalStateException("Token not stored");
        }
        if (!jwtUtil.isTokenValid(token)) {
            responseStore.setResponse(401, Map.of("message", "Unauthorized"));
            return;
        }
        String username = jwtUtil.extractUsername(token);
        userRepository.findByUsername(username).ifPresentOrElse(user -> {
            user.setStatus("DISABLED");
            userRepository.save(user);
            responseStore.setResponse(200);
        }, () -> responseStore.setResponse(404, Map.of("message", "User not found")));
    }

    @Given("^alice has deactivated her own account via POST /api/v1/users/me/deactivate$")
    public void aliceHasDeactivatedHerOwnAccount() {
        String token = tokenStore.getToken();
        if (token == null) {
            throw new IllegalStateException("Token not stored");
        }
        if (!jwtUtil.isTokenValid(token)) {
            throw new RuntimeException("Invalid token for account deactivation");
        }
        String username = jwtUtil.extractUsername(token);
        userRepository.findByUsername(username).ifPresent(user -> {
            user.setStatus("DISABLED");
            userRepository.save(user);
        });
    }

    // ============================================================
    // Internal helpers
    // ============================================================

    private void performGetUsersMe(final String token) {
        if (!jwtUtil.isTokenValid(token)) {
            responseStore.setResponse(401, Map.of("message", "Unauthorized"));
            return;
        }
        String username = jwtUtil.extractUsername(token);
        userRepository.findByUsername(username).ifPresentOrElse(
                user -> responseStore.setResponse(200, UserProfileResponse.from(user)),
                () -> responseStore.setResponse(404, Map.of("message", "User not found")));
    }

    private void performChangePassword(
            final String token, final String oldPassword, final String newPassword) {
        if (!jwtUtil.isTokenValid(token)) {
            responseStore.setResponse(401, Map.of("message", "Unauthorized"));
            return;
        }
        String username = jwtUtil.extractUsername(token);
        userRepository.findByUsername(username).ifPresentOrElse(user -> {
            if (!passwordEncoder.matches(oldPassword, user.getPasswordHash())) {
                responseStore.setResponse(401, Map.of("message", "Invalid credentials"));
                return;
            }
            user.setPasswordHash(Objects.requireNonNull(passwordEncoder.encode(newPassword)));
            userRepository.save(user);
            responseStore.setResponse(200);
        }, () -> responseStore.setResponse(404, Map.of("message", "User not found")));
    }
}

File: apps/a-demo-be-rust-axum/tests/integration/steps/user_account_steps.rs
use cucumber::{given, when};

use crate::world::AppWorld;

#[when("alice sends GET /api/v1/users/me")]
async fn get_alice_profile(world: &mut AppWorld) {
    let bearer = world.bearer();
    world.svc_get_profile(&bearer).await;
}

#[when(
    regex = r#"alice sends PATCH /api/v1/users/me with body \{ "displayName": "Alice Smith" \}"#
)]
async fn patch_display_name(world: &mut AppWorld) {
    let bearer = world.bearer();
    world.svc_update_profile(&bearer, "Alice Smith").await;
}

#[when(
    regex = r#"alice sends POST /api/v1/users/me/password with body \{ "oldPassword": "Str0ng#Pass1", "newPassword": "NewPass#456" \}"#
)]
async fn change_password_correct(world: &mut AppWorld) {
    let bearer = world.bearer();
    world
        .svc_change_password(&bearer, "Str0ng#Pass1", "NewPass#456")
        .await;
}

#[when(
    regex = r#"alice sends POST /api/v1/users/me/password with body \{ "oldPassword": "Wr0ngOld!", "newPassword": "NewPass#456" \}"#
)]
async fn change_password_wrong_old(world: &mut AppWorld) {
    let bearer = world.bearer();
    world
        .svc_change_password(&bearer, "Wr0ngOld!", "NewPass#456")
        .await;
}

#[when("alice sends POST /api/v1/users/me/deactivate")]
async fn deactivate_alice(world: &mut AppWorld) {
    let bearer = world.bearer();
    world.svc_deactivate(&bearer).await;
}

#[given("alice has deactivated her own account via POST /api/v1/users/me/deactivate")]
async fn alice_deactivated_herself(world: &mut AppWorld) {
    let bearer = world.bearer();
    world.svc_deactivate(&bearer).await;
}

#[when("the client sends GET /api/v1/users/me with alice's access token")]
async fn get_me_with_alice_token(world: &mut AppWorld) {
    let token = world.auth_token.clone().unwrap_or_default();
    world.svc_get_profile(&format!("Bearer {token}")).await;
}
